From b175b855fbc1dd67fcf275d17dcac2aa35980b91 Mon Sep 17 00:00:00 2001
From: Joerg Bornemann <joerg.bornemann@qt.io>
Date: Tue, 9 Sep 2025 13:40:02 +0200
Subject: [PATCH 063/553] CMake: Simplify the installation of .prl files

Since the commit that introduced .prl file generation,
bfcf36d459c7a36a92dd6f9127e6513f7a08277e, we jumped through some hoops
to collect the build directories containing .prl files and then
installing these directories with the FILES_MATCHING argument. This
resulted in the installation of empty directories, because
install(DIRECTORY ... FILES_MATCHING ...) preserves the input directory
structure.

This complex installation procedure is unnecessary. We know how to
construct the final .prl file path with generator expressions.
install(FILES) supports generator expressions since CMake 3.0, which the
original author of bfcf36d459c7a36a92dd6f9127e6513f7a08277e missed.

Remove the qt_internal_install_prl_files function that called
install(DIRECTORY ... FILES_MATCHING ...) in a post-process step, and
call install(FILES ...) instead for every generated .prl file. This
solves the issue that we would install empty directories in lib/cmake.

Task-number: QTBUG-138580
Change-Id: I4af3e808dc420d08a0e1a1fe3059a0a432949586
Reviewed-by: Alexandru Croitor <alexandru.croitor@qt.io>
(cherry picked from commit a975781f5c86981048b0767ae9bbe5d1e59e7160)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 cmake/QtPostProcess.cmake        |  1 -
 cmake/QtPostProcessHelpers.cmake | 16 ----------------
 cmake/QtPrlHelpers.cmake         | 11 ++++-------
 3 files changed, 4 insertions(+), 24 deletions(-)

diff --git a/cmake/QtPostProcess.cmake b/cmake/QtPostProcess.cmake
index f69448c14a4..37716460d08 100644
--- a/cmake/QtPostProcess.cmake
+++ b/cmake/QtPostProcess.cmake
@@ -13,5 +13,4 @@ if (ANDROID)
     qt_modules_process_android_dependencies()
 endif()
 
-qt_internal_install_prl_files()
 qt_internal_generate_user_facing_tools_info()
diff --git a/cmake/QtPostProcessHelpers.cmake b/cmake/QtPostProcessHelpers.cmake
index 115e2f8b585..d9fd0c873af 100644
--- a/cmake/QtPostProcessHelpers.cmake
+++ b/cmake/QtPostProcessHelpers.cmake
@@ -861,22 +861,6 @@ function(qt_internal_create_config_file_for_standalone_tests)
     )
 endfunction()
 
-function(qt_internal_install_prl_files)
-    # Get locations relative to QT_BUILD_DIR from which prl files should be installed.
-    get_property(prl_install_dirs GLOBAL PROPERTY QT_PRL_INSTALL_DIRS)
-
-    # Clear the list of install dirs so the previous values don't pollute the list of install dirs
-    # for the next repository in a top-level build.
-    set_property(GLOBAL PROPERTY QT_PRL_INSTALL_DIRS "")
-
-    foreach(prl_install_dir ${prl_install_dirs})
-        qt_install(DIRECTORY "${QT_BUILD_DIR}/${prl_install_dir}/"
-            DESTINATION ${prl_install_dir}
-            FILES_MATCHING PATTERN "*.prl"
-        )
-    endforeach()
-endfunction()
-
 function(qt_internal_generate_user_facing_tools_info)
     if("${INSTALL_PUBLICBINDIR}" STREQUAL "")
         return()
diff --git a/cmake/QtPrlHelpers.cmake b/cmake/QtPrlHelpers.cmake
index bc76a472e27..a6a5efa52f7 100644
--- a/cmake/QtPrlHelpers.cmake
+++ b/cmake/QtPrlHelpers.cmake
@@ -213,12 +213,9 @@ ${prl_step1_content_libs}
         target_sources(${target} PRIVATE "${prl_step2_path}")
     endforeach()
 
-    # Installation of the .prl file happens globally elsewhere,
-    # because we have no clue here what the actual file name is.
-    # What we know however, is the directory where the prl file is created.
-    # Save that for later, to install all prl files from that directory.
-    get_property(prl_install_dirs GLOBAL PROPERTY QT_PRL_INSTALL_DIRS)
-    if(NOT install_dir IN_LIST prl_install_dirs)
-        set_property(GLOBAL APPEND PROPERTY QT_PRL_INSTALL_DIRS "${install_dir}")
+    # Install the final .prl file that's generated by the QtFinishPrlFile.cmake script.
+    # For Apple frameworks, the .prl file is already placed inside the framework.
+    if(NOT is_framework)
+        qt_install(FILES "${final_prl_file_path}" DESTINATION "${install_dir}")
     endif()
 endfunction()
-- 
2.51.2

